[torch] Implement aten::native_batch_norm.out for CPU

larryliu0820 opened the pull request:
Summary:
Implement `native_batch_norm.out` for CPU. Reuses the main logic for `native_batch_norm` but extract out the Tensor creation logic for outputs. There are 3 outputs: `output`, `save_mean` and `save_var`. `batch_norm_cpu` calls `batch_norm_cpu_update_stats_template` to get `save_mean` and `save_var`, and then calls into `batch_norm_cpu_transform_input_template` which initializes `output`.

In the implementation of `batch_norm_cpu_out`, I did the following:

* Let `batch_norm_cpu_transform_input_template` to take another argument `output`, ask the call sites to pass in a output Tensor.

* Overload `batch_norm_cpu_update_stats_template` to take `save_mean` and `save_var`, ask the call sites to pass in those Tensors.

* In `batch_norm_cpu_out`, pass `output`, `save_mean` and `save_var` all the way to our new `batch_norm_cpu_transform_input_template` and `batch_norm_cpu_update_stats_template`.

* In `batch_norm_cpu`, prepare for these outputs and call `batch_norm_cpu_out`.

Test Plan: Enable unit tests for `native_batch_norm.out`.

Differential Revision: D40992036

iseeyuan reviewed (approved):
Nice! Thanks @larryliu0820 !

jjsjann123 commented on aten/src/ATen/native/Normalization.cpp:
```diff
@@ -123,17 +123,16 @@ std::tuple<Tensor,Tensor,Tensor> batch_norm_cpu_transform_input_template(
-    bool train, double eps) {
+    bool train, double eps, Tensor& output) {
 
   bool all_contiguous = is_contiguous(input)
```
shouldn't `all_contiguous` also checks contiguity on outputs then?

jjsjann123 commented on aten/src/ATen/native/Normalization.cpp:
```diff
@@ -690,40 +694,87 @@ std::tuple<Tensor, Tensor> batch_norm_update_stats_cpu(
+  const int64_t ndim = self.dim();
+  DimVector reduce_dims(ndim - 1);
+  reduce_dims[0] = 0;
+  for (const auto i : c10::irange(2, ndim)) {
```
nitpick `std::iota` maybe?

jjsjann123 reviewed (approved):
LGTM, minor comments

larryliu0820 committed:
[torch] Implement aten::native_batch_norm.out for CPU (#88604)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/88604

Implement `native_batch_norm.out` for CPU. Reuses the main logic for `native_batch_norm` but extract out the Tensor creation logic for outputs. There are 3 outputs: `output`, `save_mean` and `save_var`. `batch_norm_cpu` calls `batch_norm_cpu_update_stats_template` to get `save_mean` and `save_var`, and then calls into `batch_norm_cpu_transform_input_template` which initializes `output`.

In the implementation of `batch_norm_cpu_out`, I did the following:

* Let `batch_norm_cpu_transform_input_template` to take another argument `output`, ask the call sites to pass in a output Tensor.

* Overload `batch_norm_cpu_update_stats_template` to take `save_mean` and `save_var`, ask the call sites to pass in those Tensors.

* In `batch_norm_cpu_out`, pass `output`, `save_mean` and `save_var` all the way to our new `batch_norm_cpu_transform_input_template` and `batch_norm_cpu_update_stats_template`.

* In `batch_norm_cpu`, prepare for these outputs and call `batch_norm_cpu_out`.

Test Plan: Enable unit tests for `native_batch_norm.out`.

Reviewed By: iseeyuan

Differential Revision: D40992036

fbshipit-source-id: 9aa2a66abc39b171ce2ddb9c5ed4a7d6e5575a9e
